consensus/consensus_v2.go

```diff
@@ -205,6 +211,102 @@ func (consensus *Consensus) onAnnounce(msg *msg_pb.Message) {
 				Msg("[OnAnnounce] Block content is not verified successfully")
 			return
 		}
+
+		if consensus.ShardID == 0 {
+			//TODO: HB fix this
+			if recvMsg.BlockNum%core.ShardingSchedule.BlocksPerEpoch() == 0 {
+				consensus.generateNewVrf = true
+				consensus.generateNewVdf = true
+
+				consensus.vrfBlockNumbers = nil
+				consensus.ChainReader.WriteEpochVrfBlockNums(headerObj.Epoch, consensus.vrfBlockNumbers)
+			}
+
+			zeroVrfBytes := make([]byte, 32)
+			if !bytes.Equal(headerObj.Vrf[:], zeroVrfBytes) {
+				vrfPk := vrf_bls.NewVRFVerifier(consensus.LeaderPubKey)
+
+				var blockHash [32]byte
+				previousHeader := consensus.ChainReader.GetHeaderByNumber(headerObj.Number.Uint64() - 1)
+				previousHash := previousHeader.Hash()
+				copy(blockHash[:], previousHash[:])
+
+				hash, err := vrfPk.ProofToHash(blockHash[:], headerObj.VrfProof[:])
+				if err != nil {
+					consensus.getLogger().Warn().
+						Err(err).
+						Str("MsgBlockNum", headerObj.Number.String()).
+						Msg("[OnAnnounce] VRF verification error")
+					return
+				}
+
+				if hash != headerObj.Vrf {
+					consensus.getLogger().Warn().
+						Str("MsgBlockNum", headerObj.Number.String()).
+						Msg("[OnAnnounce] VRF proof is not valid")
+					return
+				}
+
+				consensus.vrfBlockNumbers, _ = consensus.ChainReader.ReadEpochVrfBlockNums(headerObj.Epoch)
+				consensus.vrfBlockNumbers = append(consensus.vrfBlockNumbers, headerObj.Number.Uint64())
+				err = consensus.ChainReader.WriteEpochVrfBlockNums(headerObj.Epoch, consensus.vrfBlockNumbers)
+				if err != nil {
+					consensus.getLogger().Error().Err(err).
+						Str("MsgBlockNum", headerObj.Number.String()).
+						Msg("[OnAnnounce] failed to write VRF block number to local db")
+				}
+
+				consensus.getLogger().Info().
+					Str("MsgBlockNum", headerObj.Number.String()).
+					Int("Number of VRF in this epoch", len(consensus.vrfBlockNumbers)).
+					Msg("[OnAnnounce] validated a new VRF")
+			}
+
+			zeroVdfBytes := make([]byte, 258)
```

Comment:
I thought vdf is 516 bytes?

Reply:
258 is VDF and another 258 is proof for the VDF.